packages/techdocs-common/src/stages/publish/awsS3.ts

```diff
@@ -0,0 +1,226 @@
+import { Logger } from 'winston';
+import { Entity, EntityName } from '@backstage/catalog-model';
+import { Config } from '@backstage/config';
+import { getHeadersForFileExtension, getFileTreeRecursively } from './helpers';
+import { PublisherBase, PublishRequest } from './types';
+import { ManagedUpload } from 'aws-sdk/clients/s3';
+import fs from 'fs';
+
+export class AwsS3Publish implements PublisherBase {
+  static fromConfig(config: Config, logger: Logger): PublisherBase {
+    let region = null;
+    let accessKeyId = null;
+    let secretAccessKey = null;
+    let bucketName = '';
+    try {
+      accessKeyId = config.getString(
+        'techdocs.publisher.awsS3.credentials.accessKeyId',
+      );
+      secretAccessKey = config.getString(
+        'techdocs.publisher.awsS3.credentials.secretAccessKey',
+      );
+      region = config.getOptionalString('techdocs.publisher.awsS3.region');
+      bucketName = config.getString('techdocs.publisher.awsS3.bucketName');
+    } catch (error) {
+      throw new Error(
+        "Since techdocs.publisher.type is set to 'awsS3' in your app config, " +
+          'credentials and bucketName are required in techdocs.publisher.awsS3 ' +
+          'required to authenticate with AWS S3.',
+      );
+    }
+
+    const storageClient = new aws.S3({
+      credentials: { accessKeyId, secretAccessKey },
+      ...(region && { region }),
+    });
+
+    // Check if the defined bucket exists. Being able to connect means the configuration is good
+    // and the storage client will work.
+    storageClient.headBucket(
+      {
+        Bucket: bucketName,
+      },
+      err => {
+        if (err) {
+          logger.error(
+            `Could not retrieve metadata about the AWS S3 bucket ${bucketName}. ` +
+              'Make sure the AWS project and the bucket exists and the access key located at the path ' +
+              "techdocs.publisher.awsS3.credentials defined in app config has the role 'Storage Object Creator'. " +
+              'Refer to https://backstage.io/docs/features/techdocs/using-cloud-storage',
+          );
+          throw new Error(`from AWS client library: ${err.message}`);
+        } else {
+          logger.info(
+            `Successfully connected to the AWS S3 bucket ${bucketName}.`,
+          );
+        }
+      },
+    );
+
+    return new AwsS3Publish(storageClient, bucketName, logger);
+  }
+
+  constructor(
+    private readonly storageClient: aws.S3,
+    private readonly bucketName: string,
+    private readonly logger: Logger,
+  ) {
+    this.storageClient = storageClient;
+    this.bucketName = bucketName;
+    this.logger = logger;
+  }
+
+  /**
+   * Upload all the files from the generated `directory` to the S3 bucket.
+   * Directory structure used in the bucket is - entityNamespace/entityKind/entityName/index.html
+   */
+  publish({ entity, directory }: PublishRequest): Promise<void> {
+    return new Promise(async (resolve, reject) => {
```

Comment:
Prefer if we could use `try` `catch` around promises  rather than encapsulating all this logic. I think it might be a little cleaner.

`fs-extra` I think will return promises for most of the operations that you require here, or you can use `fs.promises` from the standard library instead.

From a readers POV it can be super hard to read and understand without `async/await`